lib/network_utils.ml

```diff
@@ -208,6 +230,72 @@ module Sysfs = struct
 		|> (fun p -> try read_one_line p |> duplex_of_string with _ -> Duplex_unknown)
 		in (speed, duplex)
 
+	let get_dev_nums_with_same_driver driver = 
+		try
+			Sys.readdir ("/sys/bus/pci/drivers/" ^ driver)
+			|> Array.to_list
+			|> List.filter (Re.execp (Re_perl.compile_pat "\d+:\d+:\d+\.\d+"))
+			|> List.length
+		with _ -> 0
+
+	let parent_device_of_vf pcibuspath =
+		try
+			let pf_net_path = Printf.sprintf "/sys/bus/pci/devices/%s/physfn/net" pcibuspath in
+			let devices = Sys.readdir pf_net_path in
+			Result.Ok devices.(0)
+		with _ -> Result.Error (Parent_dev_for_vf_not_found, "Can not get parent device for " ^ pcibuspath)
+
+	let device_index_of_vf parent_device pcibuspath =
+		try
+			let re = Re_perl.compile_pat "virtfn(\d+)" in
+			let device_path = getpath parent_device "device" in
+			let group = Sys.readdir device_path
+				|> Array.to_list
+				|> List.filter (Re.execp re) (* List elements are like "virtfn1" *)
+				|> List.find (fun x -> Xstringext.String.has_substr (Unix.readlink (device_path ^ "/" ^ x)) pcibuspath )
+				|> Re.exec_opt re
+			in
+			match group with
+			| None -> Result.Error (Vf_index_not_found, "Can not get device index for " ^ pcibuspath)
+			| Some x -> Ok (int_of_string (Re.Group.get x 1))
+		with _ -> Result.Error (Vf_index_not_found, "Can not get device index for " ^ pcibuspath)
+
+	let get_sriov_numvfs dev =
+		try
+			getpath dev "device/sriov_numvfs"
+			|> read_one_line  
+			|> String.trim
+			|> int_of_string
+		with _ -> 0
+
+	let get_sriov_maxvfs dev =
+		try
+			getpath dev "device/sriov_totalvfs"
+			|> read_one_line  
+			|> String.trim
+			|> int_of_string
+			|> (-) 1 (* maxvfs is totalvfs -1, as totalvfs is PF num + VF num *)
```

Comment:
This computes `1 - sriov_totalvfs`. You probably want `sriov_totalvfs - 1`. Use `|> fun n -> n - 1`